controllers/configurationpolicy_controller.go

```diff
@@ -364,6 +361,104 @@ func (r *ConfigurationPolicyReconciler) handleObjectTemplates(plc policyv1.Confi
 	relatedObjects := []policyv1.RelatedObject{}
 	parentStatusUpdateNeeded := false
 
+	// Collect template information for processing:
+	// - Object name, if specified
+	// - Object namespace, if specified
+	// - Whether object is namespaced
+	//
+	// Also determine whether a namespace query is needed for this policy:
+	//   If a namespaceSelector is present AND objects are namespaced without a namespace specified
+	templateObjs := make([]struct {
```

Comment:
Optional, consider moving this new block of code to a separate function that returns `templateObjs`. That would make it easier to read and potentially unit testable if that's needed.

Reply:
Good idea!

Reply:
I might chalk this up to technical debt. I started migrating it, but I found `GetDetails` which pretty much does the same thing but differenter and then wondering if I could combine them. There are already a bunch of functions in this gigantic file but moving the code to a different package would require passing the reconciler object over to it for the `isObjectNamespaced` function, which seemed heavy for a util function. I might poke at it tomorrow when I'm clearer headed, but in this case we might need to prioritize getting the code in so QE can start testing it and consider this again when ConfigurationPolicy is (hopefully!) refactored. Or is this simpler than I'm making it out to be?

Reply:
@dhaiducek I was thinking something like this (untested and undocumented and unlinted):
```diff
diff --git a/controllers/configurationpolicy_controller.go b/controllers/configurationpolicy_controller.go
index 949fb65..69bac0b 100644
--- a/controllers/configurationpolicy_controller.go
+++ b/controllers/configurationpolicy_controller.go
@@ -335,32 +335,13 @@ func getTemplateConfigErrorMsg(err error) string {
        return fmt.Sprintf("An unexpected error occurred when configuring the template resolver: %v", err)
 }
 
-// handleObjectTemplates iterates through all policy templates in a given policy and processes them
-func (r *ConfigurationPolicyReconciler) handleObjectTemplates(plc policyv1.ConfigurationPolicy) {
-       log := log.WithValues("policy", plc.GetName())
-       log.V(1).Info("Processing object templates")
-
-       // error if no remediationAction is specified
-       if plc.Spec.RemediationAction == "" {
-               message := "Policy does not have a RemediationAction specified"
-               log.Info(message)
-               statusChanged := addConditionToStatus(&plc, 0, false, "No RemediationAction", message)
-
-               if statusChanged {
-                       r.Recorder.Event(&plc, eventWarning,
-                               fmt.Sprintf(plcFmtStr, plc.GetName()), convertPolicyStatusToString(&plc))
-               }
-
-               r.addForUpdate(&plc, statusChanged)
-
-               return
-       }
-
-       // initialize the RelatedObjects for this Configuration Policy
-       oldRelated := append([]policyv1.RelatedObject{}, plc.Status.RelatedObjects...)
-       relatedObjects := []policyv1.RelatedObject{}
-       parentStatusUpdateNeeded := false
+type objectTemplateDetails struct {
+       name         string
+       namespace    string
+       isNamespaced bool
+}
 
+func (r *ConfigurationPolicyReconciler) getObjectTemplateDetails(plc policyv1.ConfigurationPolicy) ([]objectTemplateDetails, []string, bool, error) {
        // Collect template information for processing:
        // - Object name, if specified
        // - Object namespace, if specified
@@ -368,11 +349,7 @@ func (r *ConfigurationPolicyReconciler) handleObjectTemplates(plc policyv1.Confi
        //
        // Also determine whether a namespace query is needed for this policy:
        //   If a namespaceSelector is present AND objects are namespaced without a namespace specified
-       templateObjs := make([]struct {
-               name         string
-               namespace    string
-               isNamespaced bool
-       }, len(plc.Spec.ObjectTemplates))
+       templateObjs := make([]objectTemplateDetails, len(plc.Spec.ObjectTemplates))
        selectedNamespaces := []string{}
        queryNamespaces := false
 
@@ -381,7 +358,7 @@ func (r *ConfigurationPolicyReconciler) handleObjectTemplates(plc policyv1.Confi
                if jsonErr := json.Unmarshal(objectT.ObjectDefinition.Raw, &blob); jsonErr != nil {
                        log.Error(jsonErr, "Could not unmarshal data from JSON")
 
-                       return
+                       return templateObjs, selectedNamespaces, false, jsonErr
                }
 
                var unstruct unstructured.Unstructured
@@ -436,8 +413,6 @@ func (r *ConfigurationPolicyReconciler) handleObjectTemplates(plc policyv1.Confi
                                        "Error filtering namespaces with provided NamespaceSelector: %s", err.Error())
                                statusChanged := addConditionToStatus(&plc, 0, false, reason, msg)
                                if statusChanged {
-                                       parentStatusUpdateNeeded = true
-
                                        r.Recorder.Event(
                                                &plc,
                                                eventWarning,
@@ -448,7 +423,7 @@ func (r *ConfigurationPolicyReconciler) handleObjectTemplates(plc policyv1.Confi
 
                                r.addForUpdate(&plc, statusChanged)
 
-                               return
+                               return templateObjs, selectedNamespaces, true, err
                        }
 
                        if len(selectedNamespaces) == 0 {
@@ -458,6 +433,44 @@ func (r *ConfigurationPolicyReconciler) handleObjectTemplates(plc policyv1.Confi
                }
        }
 
+       return templateObjs, selectedNamespaces, false, nil
+}
+
+// handleObjectTemplates iterates through all policy templates in a given policy and processes them
+func (r *ConfigurationPolicyReconciler) handleObjectTemplates(plc policyv1.ConfigurationPolicy) {
+       log := log.WithValues("policy", plc.GetName())
+       log.V(1).Info("Processing object templates")
+
+       // error if no remediationAction is specified
+       if plc.Spec.RemediationAction == "" {
+               message := "Policy does not have a RemediationAction specified"
+               log.Info(message)
+               statusChanged := addConditionToStatus(&plc, 0, false, "No RemediationAction", message)
+
+               if statusChanged {
+                       r.Recorder.Event(&plc, eventWarning,
+                               fmt.Sprintf(plcFmtStr, plc.GetName()), convertPolicyStatusToString(&plc))
+               }
+
+               r.addForUpdate(&plc, statusChanged)
+
+               return
+       }
+
+       // initialize the RelatedObjects for this Configuration Policy
+       oldRelated := append([]policyv1.RelatedObject{}, plc.Status.RelatedObjects...)
+       relatedObjects := []policyv1.RelatedObject{}
+       parentStatusUpdateNeeded := false
+
+       templateObjs, selectedNamespaces, parentStatusUpdateNeeded, err := r.getObjectTemplateDetails(plc)
+       if err != nil {
+               if parentStatusUpdateNeeded {
+                       r.checkRelatedAndUpdate(plc, relatedObjects, oldRelated, parentStatusUpdateNeeded)
+               }
+
+               return
+       }
+
        addTemplateErrorViolation := func(reason, msg string) {
                log.Info("Setting the policy to noncompliant due to a templating error", "error", msg)
```

Reply:
Yeah, yesterday I was trying to get it the heck out of the giant file, but this at least does make it more readable. I'll update it--thanks!

Reply:
@dhaiducek you could move it to another file in the same package if you want though.